File: source-googlesheet/src/main/java/ru/pocketbyte/locolaser/source/google/parser/GoogleSheetConfigParser.java
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.source.google.parser;

import org.json.simple.JSONObject;
import ru.pocketbyte.locolaser.config.parser.BaseTableSourceConfigParser;
import ru.pocketbyte.locolaser.exception.InvalidConfigException;
import ru.pocketbyte.locolaser.source.google.sheet.GoogleSheetConfig;
import ru.pocketbyte.locolaser.config.parser.ConfigParser;
import ru.pocketbyte.locolaser.utils.JsonParseUtils;
import ru.pocketbyte.locolaser.utils.TextUtils;

import java.net.MalformedURLException;

/**
 * @author Denis Shurygin
 */
public class GoogleSheetConfigParser extends BaseTableSourceConfigParser<GoogleSheetConfig> {

    public static final String SHEET_ID = "id";
    public static final String SHEET_WORKSHEET_TITLE = "worksheet_title";
    public static final String SHEET_CREDENTIAL_FILE = "credential_file";

    protected GoogleSheetConfig sourceByType(String type) throws InvalidConfigException {
        if (type != null && !type.equals(GoogleSheetConfig.TYPE))
            throw new InvalidConfigException("Source type is \"" + type + "\", but expected \"" + GoogleSheetConfig.TYPE + "\".");

        return new GoogleSheetConfig();
    }

    protected void fillFromJSON(GoogleSheetConfig source, JSONObject configJson) throws InvalidConfigException {
        super.fillFromJSON(source, configJson);
        source.setId(JsonParseUtils.getString(configJson, SHEET_ID, ConfigParser.SOURCE, true));

        source.setWorksheetTitle(JsonParseUtils.getString(
                                        configJson, SHEET_WORKSHEET_TITLE, ConfigParser.SOURCE, false));

        source.setCredentialFile(JsonParseUtils.getFile(
                                        configJson, SHEET_CREDENTIAL_FILE, ConfigParser.SOURCE, false));
    }

    @Override
    protected void validate(GoogleSheetConfig source) throws InvalidConfigException {
        super.validate(source);

        if (TextUtils.isEmpty(source.getId()))
            throw new InvalidConfigException("\"" + ConfigParser.SOURCE + "." + SHEET_ID + "\" is not set.");
        try {
            source.getUrl();
        } catch (MalformedURLException e) {
            throw new InvalidConfigException("\"" + ConfigParser.SOURCE + "." + SHEET_ID + "\" is invalid.");
        }
    }
}

File: source-googlesheet/src/main/java/ru/pocketbyte/locolaser/source/google/parser/GoogleSheetConfigParser.kt
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.source.google.parser

import org.json.simple.JSONObject
import ru.pocketbyte.locolaser.config.parser.BaseTableSourceConfigParser
import ru.pocketbyte.locolaser.config.parser.ConfigParser.Companion.SOURCE
import ru.pocketbyte.locolaser.exception.InvalidConfigException
import ru.pocketbyte.locolaser.source.google.sheet.GoogleSheetConfig
import ru.pocketbyte.locolaser.utils.json.JsonParseUtils
import ru.pocketbyte.locolaser.utils.json.JsonParseUtils.getString

/**
 * @author Denis Shurygin
 */
class GoogleSheetConfigParser : BaseTableSourceConfigParser<GoogleSheetConfig>() {

    companion object {
        const val SHEET_ID = "id"
        const val SHEET_WORKSHEET_TITLE = "worksheet_title"
        const val SHEET_CREDENTIAL_FILE = "credential_file"
    }

    @Throws(InvalidConfigException::class)
    override fun sourceByType(type: String?): GoogleSheetConfig {
        if (type != null && type != GoogleSheetConfig.TYPE)
            throw InvalidConfigException("Source type is \"$type\", but expected \"${GoogleSheetConfig.TYPE}\".")

        return GoogleSheetConfig()
    }

    @Throws(InvalidConfigException::class)
    override fun fillFromJSON(source: GoogleSheetConfig, configJson: JSONObject) {
        super.fillFromJSON(source, configJson)
        source.id = getString(configJson, SHEET_ID, SOURCE, true)

        source.worksheetTitle = getString(
                configJson, SHEET_WORKSHEET_TITLE, SOURCE, false)

        source.credentialFile = JsonParseUtils.getFile(
                configJson, SHEET_CREDENTIAL_FILE, SOURCE, false)
    }

    @Throws(InvalidConfigException::class)
    override fun validate(source: GoogleSheetConfig) {
        super.validate(source)

        if (source.id?.isEmpty() != false)
            throw InvalidConfigException("\"$SOURCE.$SHEET_ID\" is not set.")

    }
}
